Hookup mime classifier to main document loader

TimvdLippe opened the pull request:
While we don't perform any sniffing yet, it does now
check all relevant mime types. Before, we would only
check specific versions of specific mime types. However,
the specification is more elaborate, which MimeClassifier
already handles.

There are two new test failures, but they actually fail
in the same way in other browsers. These tests are still
being discussed what the correct behavior should be [1]

Part of #14024

[1]: https://github.com/whatwg/mimesniff/issues/189#issuecomment-2081559661

TimvdLippe committed:
Add spec comments to MimeClassifier

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Reduce nesting in MimeClassifier::classify

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Hookup mime classifier to main document loader

While we don't perform any sniffing yet, it does now
check all relevant mime types. Before, we would only
check specific versions of specific mime types. However,
the specification is more elaborate, which MimeClassifier
already handles.

There are two new test failures, but they actually fail
in the same way in other browsers. These tests are still
being discussed what the correct behavior should be [1]

Part of #14024

[1]: https://github.com/whatwg/mimesniff/issues/189#issuecomment-2081559661

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

jdm commented on components/script/dom/servoparser/mod.rs:
```diff
@@ -943,49 +946,70 @@ impl FetchResponseListener for ParserContext {
+            &content_type,
+            // TODO(14024): Figure out how to pass the response data here for sniffing
+            // Requires implementation of https://mimesniff.spec.whatwg.org/#read-the-resource-header
+            &[],
```
We're going to need to move a bunch of this logic out of process_response and into process_response_chunk to support this. Fun!

TimvdLippe replied:
Yup. Will do that in a follow-up to keep things manageable.

jdm reviewed (approved):
This looks pretty reasonable!

TimvdLippe committed:
Remove feed sniffing

This was removed from the standard in
https://github.com/whatwg/mimesniff/commit/b47be44aba066f3ef03864d9ed24eb8c7b533758

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
Also had to add https://github.com/servo/servo/pull/39109/commits/57c7ad0bc7536788bf55c47813873597ea776e22 to fix the unit tests. Turns out that functionality is no longer present in the spec (see https://github.com/whatwg/mimesniff/commit/b47be44aba066f3ef03864d9ed24eb8c7b533758)
